newIDE/app/src/EventsFunctionsList/index.js

```diff
@@ -6,509 +6,1291 @@ import { t } from '@lingui/macro';
+export const extensionFunctionsRootFolderId = 'extension-functions';
+const extensionObjectsEmptyPlaceholderId = 'extension-objects-placeholder';
+const extensionBehaviorsEmptyPlaceholderId = 'extension-behaviors-placeholder';
+const extensionFunctionsEmptyPlaceholderId = 'extension-functions-placeholder';
+
 const styles = {
   listContainer: {
     flex: 1,
+    display: 'flex',
+    flexDirection: 'column',
   },
-  tooltip: { marginRight: 5, verticalAlign: 'bottom' },
+  autoSizerContainer: { flex: 1 },
+  autoSizer: { width: '100%' },
 };
 
-export type EventsFunctionCreationParameters = {|
-  functionType: 0 | 1 | 2,
-  name: ?string,
+const extensionItemReactDndType = 'GD_EXTENSION_ITEM';
+
+export interface TreeViewItemContent {
+  getName(): string | React.Node;
+  getId(): string;
+  getHtmlId(index: number): ?string;
+  getThumbnail(): ?string;
+  getDataset(): ?HTMLDataset;
+  onSelect(): void;
+  buildMenuTemplate(i18n: I18nType, index: number): Array<MenuItemTemplate>;
+  getRightButton(): ?MenuButton;
+  renderLeftComponent(i18n: I18nType): ?React.Node;
+  rename(newName: string): void;
+  edit(): void;
+  delete(): void;
+  getIndex(): number;
+  moveAt(destinationIndex: number): void;
+  isDescendantOf(itemContent: TreeViewItemContent): boolean;
+  getEventsFunctionsContainer(): ?gdEventsFunctionsContainer;
+  getFunctionInsertionIndex(): number;
+  getEventsFunction(): ?gdEventsFunction;
+  getEventsBasedBehavior(): ?gdEventsBasedBehavior;
+  getEventsBasedObject(): ?gdEventsBasedObject;
+}
+
+interface TreeViewItem {
+  isRoot?: boolean;
+  isPlaceholder?: boolean;
+  +content: TreeViewItemContent;
+  getChildren(i18n: I18nType): ?Array<TreeViewItem>;
+}
+
+export type TreeItemProps = {|
+  forceUpdate: () => void,
+  forceUpdateList: () => void,
+  unsavedChanges?: ?UnsavedChanges,
+  project: gdProject,
+  eventsFunctionsExtension: gdEventsFunctionsExtension,
+  editName: (itemId: string) => void,
+  scrollToItem: (itemId: string) => void,
 |};
 
-const renderEventsFunctionLabel = (eventsFunction: gdEventsFunction) => {
-  const label = (
-    <Text noMargin size="body-small">
-      {eventsFunction.getName()}
-    </Text>
-  );
-
-  return eventsFunction.isPrivate() ? (
-    <>
-      <Tooltip
-        title={
-          <Trans>This function won't be visible in the events editor.</Trans>
-        }
-      >
-        <VisibilityOff fontSize="small" style={styles.tooltip} />
-      </Tooltip>
-      {label}
-    </>
-  ) : eventsFunction.isAsync() ? (
-    <>
-      <Tooltip
-        title={
-          <Trans>
-            This function is asynchronous - it will only allow subsequent events
-            to run after calling the action "End asynchronous task" within the
-            function.
-          </Trans>
-        }
-      >
-        <AsyncIcon fontSize="small" style={styles.tooltip} />
-      </Tooltip>
-      {label}
-    </>
-  ) : (
-    label
-  );
-};
+class ObjectTreeViewItem implements TreeViewItem {
```

Comment:
Should it be named `ExtensionTreeViewItem`?

Reply:
Oh ok, it's for a CustomObject in fact

Reply:
It's for event-based objects.